staging/src/k8s.io/apiserver/plugin/pkg/audit/dynamic/dynamic.go

```diff
@@ -0,0 +1,292 @@
+	recorder             record.EventRecorder
+}
+
+type syncedBackends map[types.UID]*Backend
+
+// Names returns the names of the backend configurations
+func (s syncedBackends) Names() []string {
+	names := []string{}
+	for _, backend := range s {
+		names = append(names, backend.configuration.Name)
+	}
+	return names
+}
+
+// ProcessEvents proccesses the given events per current backend map
+func (d *dynamic) ProcessEvents(events ...*auditinternal.Event) {
+	for _, b := range d.GetBackends() {
+		b.delegate.ProcessEvents(events...)
+	}
+}
+
+// Run the dynamic backends
+func (d *dynamic) Run(stopCh <-chan struct{}) error {
+	// Shutdown channels are handled individually with the dynamic
+	// backend, message must be propagated out
+	go func() {
+		<-stopCh
+		d.disableUpdates()
+		d.stopAllBackends()
+	}()
+	var funcs []func() error
+	for _, b := range d.GetBackends() {
+		funcs = append(funcs, func() error {
+			return b.delegate.Run(b.stopChan)
+		})
+	}
+	return errors.AggregateGoroutines(funcs...)
+}
+
+// disableUpdates permanently disables the ability to modify backends
+// this is used to shutdown gracefully
+func (d *dynamic) disableUpdates() {
+	d.backendUpdateMutex.Lock()
+}
+
+// stopAllBackends closes the stopChan for every backend to enable
+// goroutines to terminate gracefully
+func (d *dynamic) stopAllBackends() {
+	for _, b := range d.GetBackends() {
+		close(b.stopChan)
+	}
+}
+
+// Shutdown calls the shutdown method on all backends
+func (d *dynamic) Shutdown() {
+	for _, b := range d.GetBackends() {
+		b.delegate.Shutdown()
+	}
+}
+
+// GetBackends retrieves current backends in a safe manner
+func (d *dynamic) GetBackends() syncedBackends {
+	return d.backends.Load().(syncedBackends)
+}
+
+// setBackends sets the current backends in a safe manner
+func (d *dynamic) setBackends(backends syncedBackends) {
+	d.backends.Store(backends)
+}
+
+// addSink is called by the shared informer when a sink is added
+func (d *dynamic) addSink(sink *auditregv1alpha1.AuditSink) {
+	b, err := d.createBackend(sink)
+	if err != nil {
+		msg := fmt.Sprintf("Could not add audit sink %q: %v", sink.Name, err)
+		glog.Error(msg)
+		d.recorder.Event(sink, v1.EventTypeWarning, "CreateFailed", msg)
+		return
+	}
+	backends := d.GetBackends()
+	if _, ok := backends[sink.UID]; ok {
+		glog.Errorf("Audit sink %q uid: %s already exists, could not readd", sink.Name, sink.UID)
+		return
+	}
+	backends[sink.UID] = b
+	d.backendUpdateMutex.Lock()
+	defer d.backendUpdateMutex.Unlock()
+	d.setBackends(backends)
+	glog.V(2).Infof("Added audit sink: %s", sink.Name)
+	glog.V(2).Infof("Current audit sinks: %v", backends.Names())
+}
+
+// updateSink is called by the shared informer when a sink is updated.
+// The new sink must not have the same uid as the previous. The new
+// sink will be started before the old one is closed so no events will be lost
+func (d *dynamic) updateSink(oldSink, newSink *auditregv1alpha1.AuditSink) {
+	b, err := d.createBackend(newSink)
```

Comment:
this is counting on addSink/updateSink/deleteSink not being called asynchronously, right?

createBackend() returns a started backend. should we wait until we know we're going to store this backend before starting it? (i.e. lock, make sure the old backend exists, then start the new backend, store, unlock)?

 same question goes for addSink